File: src/gearbox_sim/simulation/result.py
"""Simulation result container."""

from dataclasses import dataclass, field
from typing import Any, Dict, List, Optional

import numpy as np
from numpy.typing import NDArray


@dataclass
class SimulationResult:
    """Container for simulation results.

    All time series are numpy arrays with shape (n_time_points,).

    Attributes:
        time: Time points [s]
        states: State variables over time {name: array}
        controls: Control inputs over time {name: array}
        outputs: Derived outputs over time {name: array}
        metadata: Additional simulation metadata
    """

    time: NDArray
    states: Dict[str, NDArray] = field(default_factory=dict)
    controls: Dict[str, NDArray] = field(default_factory=dict)
    outputs: Dict[str, NDArray] = field(default_factory=dict)
    metadata: Dict[str, Any] = field(default_factory=dict)

    @property
    def n_points(self) -> int:
        """Number of time points."""
        return len(self.time)

    @property
    def duration(self) -> float:
        """Simulation duration [s]."""
        return self.time[-1] - self.time[0]

    @property
    def dt(self) -> float:
        """Average time step [s]."""
        return self.duration / (self.n_points - 1) if self.n_points > 1 else 0.0

    def get_state(self, name: str) -> Optional[NDArray]:
        """Get a state variable by name."""
        return self.states.get(name)

    def get_control(self, name: str) -> Optional[NDArray]:
        """Get a control input by name."""
        return self.controls.get(name)

    def get_output(self, name: str) -> Optional[NDArray]:
        """Get a derived output by name."""
        return self.outputs.get(name)

    # Convenience properties for common outputs
    @property
    def velocity(self) -> Optional[NDArray]:
        """Vehicle velocity [m/s]."""
        return self.outputs.get("velocity")

    @property
    def velocity_kmh(self) -> Optional[NDArray]:
        """Vehicle velocity [km/h]."""
        v = self.velocity
        return v * 3.6 if v is not None else None

    @property
    def soc(self) -> Optional[NDArray]:
        """Battery state of charge [0-1]."""
        # Look for SOC in states (component.SOC format)
        for name, values in self.states.items():
            if name.endswith(".SOC") or name == "SOC":
                return values
        return None

    @property
    def power_engine(self) -> Optional[NDArray]:
        """Engine power [W]."""
        return self.outputs.get("P_engine")

    @property
    def fuel_rate(self) -> Optional[NDArray]:
        """Fuel consumption rate [kg/s]."""
        return self.outputs.get("fuel_rate")

    @property
    def fuel_total(self) -> Optional[float]:
        """Total fuel consumed [kg]."""
        rate = self.fuel_rate
        if rate is not None and self.n_points > 1:
            return float(np.trapezoid(rate, self.time))
        return None

    def get_final_state(self) -> Dict[str, float]:
        """Get final values of all states."""
        return {name: float(values[-1]) for name, values in self.states.items()}

    def get_max(self, name: str) -> Optional[float]:
        """Get maximum value of a variable."""
        for source in [self.states, self.controls, self.outputs]:
            if name in source:
                return float(np.max(source[name]))
        return None

    def get_min(self, name: str) -> Optional[float]:
        """Get minimum value of a variable."""
        for source in [self.states, self.controls, self.outputs]:
            if name in source:
                return float(np.min(source[name]))
        return None

    def get_mean(self, name: str) -> Optional[float]:
        """Get mean value of a variable."""
        for source in [self.states, self.controls, self.outputs]:
            if name in source:
                return float(np.mean(source[name]))
        return None

    def slice(self, t_start: float, t_end: float) -> "SimulationResult":
        """Extract a time slice of the results.

        Args:
            t_start: Start time [s]
            t_end: End time [s]

        Returns:
            New SimulationResult with sliced data
        """
        mask = (self.time >= t_start) & (self.time <= t_end)

        return SimulationResult(
            time=self.time[mask],
            states={k: v[mask] for k, v in self.states.items()},
            controls={k: v[mask] for k, v in self.controls.items()},
            outputs={k: v[mask] for k, v in self.outputs.items()},
            metadata=self.metadata.copy(),
        )

    def resample(self, dt: float) -> "SimulationResult":
        """Resample results to a new time step.

        Args:
            dt: New time step [s]

        Returns:
            New SimulationResult with resampled data
        """
        new_time = np.arange(self.time[0], self.time[-1] + dt / 2, dt)

        def interp(arr):
            return np.interp(new_time, self.time, arr)

        return SimulationResult(
            time=new_time,
            states={k: interp(v) for k, v in self.states.items()},
            controls={k: interp(v) for k, v in self.controls.items()},
            outputs={k: interp(v) for k, v in self.outputs.items()},
            metadata=self.metadata.copy(),
        )

    def to_dataframe(self):
        """Convert to pandas DataFrame (if pandas available)."""
        try:
            import pandas as pd

            data = {"time": self.time}
            for prefix, source in [
                ("state.", self.states),
                ("control.", self.controls),
                ("output.", self.outputs),
            ]:
                for name, values in source.items():
                    data[prefix + name] = values

            return pd.DataFrame(data)
        except ImportError:
            raise ImportError("pandas is required for to_dataframe()")

    def summary(self) -> str:
        """Generate a text summary of results."""
        lines = [
            f"Simulation Results",
            f"  Duration: {self.duration:.1f} s ({self.n_points} points)",
            f"  States: {list(self.states.keys())}",
            f"  Controls: {list(self.controls.keys())}",
            f"  Outputs: {list(self.outputs.keys())}",
        ]

        if self.velocity is not None:
            v = self.velocity
            lines.append(
                f"  Velocity: {v[0]*3.6:.1f} -> {v[-1]*3.6:.1f} km/h "
                f"(max {np.max(v)*3.6:.1f} km/h)"
            )

        if self.soc is not None:
            soc = self.soc
            lines.append(f"  SOC: {soc[0]*100:.1f}% -> {soc[-1]*100:.1f}%")

        if self.fuel_total is not None:
            lines.append(f"  Fuel consumed: {self.fuel_total:.2f} kg")

        return "\n".join(lines)

    def __repr__(self) -> str:
        return f"SimulationResult(duration={self.duration:.1f}s, n_points={self.n_points})"

File: web/packages/drivetrain-sim/src/simulation/result.ts
/**
 * Simulation result container.
 */

import { interp } from '../math/interpolate';

/**
 * Container for simulation results.
 *
 * All time series are arrays with length n_time_points.
 */
export interface SimulationResult {
  /** Time points [s] */
  time: number[];
  /** State variables over time {name: array} */
  states: Record<string, number[]>;
  /** Control inputs over time {name: array} */
  controls: Record<string, number[]>;
  /** Derived outputs over time {name: array} */
  outputs: Record<string, number[]>;
  /** Additional simulation metadata */
  metadata: Record<string, unknown>;
}

/**
 * Create an empty simulation result.
 */
export function createResult(): SimulationResult {
  return {
    time: [],
    states: {},
    controls: {},
    outputs: {},
    metadata: {},
  };
}

/**
 * Get number of time points.
 */
export function getNumPoints(result: SimulationResult): number {
  return result.time.length;
}

/**
 * Get simulation duration [s].
 */
export function getDuration(result: SimulationResult): number {
  const n = result.time.length;
  return n > 1 ? result.time[n - 1] - result.time[0] : 0;
}

/**
 * Get average time step [s].
 */
export function getAverageDt(result: SimulationResult): number {
  const n = result.time.length;
  return n > 1 ? getDuration(result) / (n - 1) : 0;
}

/**
 * Get a state variable by name.
 */
export function getState(result: SimulationResult, name: string): number[] | null {
  return result.states[name] ?? null;
}

/**
 * Get a control input by name.
 */
export function getControl(result: SimulationResult, name: string): number[] | null {
  return result.controls[name] ?? null;
}

/**
 * Get a derived output by name.
 */
export function getOutput(result: SimulationResult, name: string): number[] | null {
  return result.outputs[name] ?? null;
}

/**
 * Get vehicle velocity [m/s].
 */
export function getVelocity(result: SimulationResult): number[] | null {
  return result.outputs.velocity ?? null;
}

/**
 * Get vehicle velocity [km/h].
 */
export function getVelocityKmh(result: SimulationResult): number[] | null {
  const v = getVelocity(result);
  return v ? v.map((val) => val * 3.6) : null;
}

/**
 * Get battery state of charge [0-1].
 */
export function getSoc(result: SimulationResult): number[] | null {
  // Look for SOC in states (component.SOC format)
  for (const [name, values] of Object.entries(result.states)) {
    if (name.endsWith('.SOC') || name === 'SOC') {
      return values;
    }
  }
  return null;
}

/**
 * Get engine power [W].
 */
export function getEnginePower(result: SimulationResult): number[] | null {
  return result.outputs.P_engine ?? null;
}

/**
 * Get fuel consumption rate [kg/s].
 */
export function getFuelRate(result: SimulationResult): number[] | null {
  return result.outputs.fuel_rate ?? null;
}

/**
 * Get total fuel consumed [kg] using trapezoidal integration.
 */
export function getFuelTotal(result: SimulationResult): number | null {
  const rate = getFuelRate(result);
  if (!rate || result.time.length < 2) {
    return null;
  }

  // Trapezoidal integration
  let total = 0;
  for (let i = 1; i < result.time.length; i++) {
    const dt = result.time[i] - result.time[i - 1];
    total += 0.5 * (rate[i - 1] + rate[i]) * dt;
  }
  return total;
}

/**
 * Get final values of all states.
 */
export function getFinalState(result: SimulationResult): Record<string, number> {
  const final: Record<string, number> = {};
  for (const [name, values] of Object.entries(result.states)) {
    final[name] = values[values.length - 1];
  }
  return final;
}

/**
 * Get maximum value of a variable.
 */
export function getMax(result: SimulationResult, name: string): number | null {
  for (const source of [result.states, result.controls, result.outputs]) {
    if (name in source) {
      return Math.max(...source[name]);
    }
  }
  return null;
}

/**
 * Get minimum value of a variable.
 */
export function getMin(result: SimulationResult, name: string): number | null {
  for (const source of [result.states, result.controls, result.outputs]) {
    if (name in source) {
      return Math.min(...source[name]);
    }
  }
  return null;
}

/**
 * Get mean value of a variable.
 */
export function getMean(result: SimulationResult, name: string): number | null {
  for (const source of [result.states, result.controls, result.outputs]) {
    if (name in source) {
      const arr = source[name];
      return arr.reduce((a, b) => a + b, 0) / arr.length;
    }
  }
  return null;
}

/**
 * Extract a time slice of the results.
 */
export function sliceResult(
  result: SimulationResult,
  tStart: number,
  tEnd: number
): SimulationResult {
  const indices: number[] = [];
  for (let i = 0; i < result.time.length; i++) {
    if (result.time[i] >= tStart && result.time[i] <= tEnd) {
      indices.push(i);
    }
  }

  const sliceArray = (arr: number[]) => indices.map((i) => arr[i]);

  return {
    time: sliceArray(result.time),
    states: Object.fromEntries(
      Object.entries(result.states).map(([k, v]) => [k, sliceArray(v)])
    ),
    controls: Object.fromEntries(
      Object.entries(result.controls).map(([k, v]) => [k, sliceArray(v)])
    ),
    outputs: Object.fromEntries(
      Object.entries(result.outputs).map(([k, v]) => [k, sliceArray(v)])
    ),
    metadata: { ...result.metadata },
  };
}

/**
 * Resample results to a new time step.
 */
export function resampleResult(result: SimulationResult, dt: number): SimulationResult {
  const newTime: number[] = [];
  for (let t = result.time[0]; t <= result.time[result.time.length - 1]; t += dt) {
    newTime.push(t);
  }

  const interpArray = (arr: number[]) =>
    newTime.map((t) => interp(t, result.time, arr));

  return {
    time: newTime,
    states: Object.fromEntries(
      Object.entries(result.states).map(([k, v]) => [k, interpArray(v)])
    ),
    controls: Object.fromEntries(
      Object.entries(result.controls).map(([k, v]) => [k, interpArray(v)])
    ),
    outputs: Object.fromEntries(
      Object.entries(result.outputs).map(([k, v]) => [k, interpArray(v)])
    ),
    metadata: { ...result.metadata },
  };
}

/**
 * Generate a text summary of results.
 */
export function summarizeResult(result: SimulationResult): string {
  const lines = [
    'Simulation Results',
    `  Duration: ${getDuration(result).toFixed(1)} s (${getNumPoints(result)} points)`,
    `  States: ${Object.keys(result.states).join(', ')}`,
    `  Controls: ${Object.keys(result.controls).join(', ')}`,
    `  Outputs: ${Object.keys(result.outputs).join(', ')}`,
  ];

  const v = getVelocity(result);
  if (v) {
    const vKmh = v.map((val) => val * 3.6);
    lines.push(
      `  Velocity: ${vKmh[0].toFixed(1)} -> ${vKmh[vKmh.length - 1].toFixed(1)} km/h ` +
        `(max ${Math.max(...vKmh).toFixed(1)} km/h)`
    );
  }

  const soc = getSoc(result);
  if (soc) {
    lines.push(
      `  SOC: ${(soc[0] * 100).toFixed(1)}% -> ${(soc[soc.length - 1] * 100).toFixed(1)}%`
    );
  }

  const fuel = getFuelTotal(result);
  if (fuel !== null) {
    lines.push(`  Fuel consumed: ${fuel.toFixed(2)} kg`);
  }

  return lines.join('\n');
}
